packages/squiggle-lang/src/rescript/FunctionRegistry/Library/FR_Danger.res

```diff
@@ -0,0 +1,464 @@
+                aLambda,
+                min,
+                max,
+                (max -. min) /. epsilon,
+                env,
+                reducer,
+              )->E.R2.errMap(b =>
+                "Integration error 7 in Danger.integrate. Something went wrong along the way: " ++ b
+              )
+            | _ =>
+              Error(
+                "Integration error 8 in Danger.integrate. Remember that inputs are (function, number (min), number (max), number(increment))",
+              )
+            }
+            result
+          },
+          (),
+        ),
+      ],
+      (),
+    )
+  }
+}
+
+module DiminishingReturns = {
+  module Helpers = {
+    type diminishingReturnsAccumulatorInner = {
+      optimalAllocations: array<float>,
+      currentMarginalReturns: result<array<float>, string>,
+    }
+    let findBiggestElementIndex = xs =>
+      E.A.reducei(xs, 0, (acc, newElement, index) => {
+        switch newElement > xs[acc] {
+        | true => index
+        | false => acc
+        }
+      })
+    type diminishingReturnsAccumulator = result<diminishingReturnsAccumulatorInner, string>
+    // TODO: This is so complicated, it probably should be its own file. It might also make sense to have it work in Rescript directly, taking in a function rather than a reducer; then something else can wrap that function in the reducer/lambdas/environment.
+    /*
+    The key idea for this function is that 
+    1. we keep track of past spending and current marginal returns for each function
+    2. we look an additional increment in funds
+    3. we assign it to the function with the best marginal returns
+    4. we update the spending, and we compute the new returns for that function, with more spending
+      - But we only compute the new marginal returns for the function we end up assigning the spending to.
+    5. We continue doing this until all the funding is exhausted
+    This is currently being done with a reducer, that keeps track of:
+      - Value of marginal spending for each function
+      - How much has been assigned to each function.
+    */
+
+    /*
+      Two possible algorithms (n=funds/increment, m=num lambdas)
+      1. O(n): Iterate through value on next n dollars. At each step, only compute the new marginal return of the function which is spent. (This is what we are doing.)
+      2. O(n*(m-1)): Iterate through all possible spending combinations. The advantage of this option is that it wouldn't assume that the returns of marginal spending are diminishing.
+    */
+    let optimalAllocationGivenDiminishingMarginalReturnsForManyFunctions = (
+      lambdas,
+      funds,
+      approximateIncrement,
+      environment,
+      reducer,
+    ) => {
+
+      switch (
+        E.A.length(lambdas) > 1,
+        funds > 0.0,
+        approximateIncrement > 0.0,
+        funds > approximateIncrement,
+      ) {
+      | (false, _, _, _) =>
```

Comment:
This pattern is interesting, but I don't recommend. If statements are better for things like this, imo. (They're also recommended by Rescript style guides, where are applicable)